project-lisa/source-code/client/src/api.ts

```diff
@@ -1,24 +1,261 @@
+		)
+	}
+
+	const normalizedUrl = `https://${host}/${owner}/${repoName}`
+
+	let pending: Promise<void> | undefined = raw.clone({
+		fs: wrap(rawFs, 'clone'),
+		http,
+		dir: "/",
+		corsProxy,
+		url: normalizedUrl,
+		singleBranch: true,
+		depth: 1,
+		noTags: true,
+	}).finally(() => {
+		pending = undefined
+	}).catch((err: any) => {
+		console.error('error cloning the repository', err)
+	})
+
+	// delay all fs and repo operations until the repo clone and checkout have finished, this is preparation for the lazy feature
+	function delayedAction ({ execute }: { execute: () => any }) {
+		if (pending) {
+			return pending.then(execute)
+		}
+
+		return execute()
+	}
+
+	return {
+		fs: wrap(rawFs, 'app', delayedAction),
+
+		/**
+		 * Gets the git origin url of the current repository.
+		 *
+		 * @returns The git origin url or undefined if it could not be found.
+		 */
+		async listRemotes () {
+			try {
+				const wrappedFS = wrap(rawFs, 'listRemotes', delayedAction)
+
+				const remotes = await raw.listRemotes({
+					fs: wrappedFS,
+					dir: await raw.findRoot({
+						fs: wrappedFS,
+							filepath: "/"
+					}),
+				})
+
+				return remotes
+			} catch (_err) {
+				return undefined
+			}
+		},
+
+		status ({ dir = "/", filepath }) {
+			return raw.status({
+				fs: wrap(rawFs, 'statusMatrix', delayedAction),
+				dir,
+				filepath
+			})
+		},
+
+		statusMatrix ({ dir = "/", filter }) {
+			return raw.statusMatrix({
+				fs: wrap(rawFs, 'statusMatrix', delayedAction),
+				dir,
+				filter
+			})
+		},
+
+		add ({ dir = "/", filepath }) {
+			return raw.add({
+				fs: wrap(rawFs, 'add', delayedAction),
+				dir,
+				filepath
+			})
+		},
+
+		commit ({ dir = "/", author, message }) {
+			return raw.commit({
+				fs: wrap(rawFs, 'commit', delayedAction),
+				dir,
+				author,
+				message
+			})
+		},
+
+		push ({ dir = "/" }) {
+			return raw.push({
+				fs: wrap(rawFs, 'push', delayedAction),
+				url: normalizedUrl,
+				corsProxy,
+				http,
+				dir
+			})
+		},
+
+		pull ({ dir = "/", author, fastForward, singleBranch }) {
+			return raw.pull({
+				fs: wrap(rawFs, 'pull', delayedAction),
+				url: normalizedUrl,
+				corsProxy,
+				http,
+				dir,
+				fastForward,
+				singleBranch,
+				author
+			})
+		},
+
+		log ({ dir = "/", since }) {
+			return raw.log({
+				fs: wrap(rawFs, 'log', delayedAction),
+				dir,
+				since
+			})
+		},
+
+		mergeUpstream ({ branch }) {
+			return github.request("POST /repos/{owner}/{repo}/merge-upstream", {
+				branch,
+				owner,
+				repo: repoName,
+			})
+		},
+
+		async isCollaborator ({ username }) {
+			let response: Awaited<ReturnType<typeof github.request<"GET /repos/{owner}/{repo}/collaborators/{username}">>> | undefined
+			try {
+				response = await github.request(
+					"GET /repos/{owner}/{repo}/collaborators/{username}",
+					{
+						owner,
+						repo: repoName,
+						username,
+					},
+				)
+			} catch (_err) { /* throws on non collaborator access */ }
```

Comment:
Needs to be extracted in the future and pushed into `createGitHubAdapter()` or a similar pattern. (We don't want lock-in to a specific git host. Users are using Azure, Gitlab, Bitbucket, etc.)

Reply:
No action required. hardcoded is fine for now